### PR TITLE
Add a error check in postHijacked to avoid docker exec command blocking.

### PR DESCRIPTION
**Pull request by keloyang**
When user execute docker exec command(client.ContainerExecAttach), docker daemon maybe have err return because
of ExecExists check, and then the hijack stream will not be close, it can lead to docker exec command block.

Signed-off-by: yangshukui <yangshukui@huawei.com>

**- What I did**
Add a error check in postHijacked to avoid docker exec command blocking.

**- How to verify it**

with the follown steps， the problem can be reproduced easyly.
1. start a testone container
```
docker run -d --name testone ubuntu bash -c 'for (( i = 1; i <= 200000; i++ )) do echo -n "$i";sleep 1;done'
```
2. run pause.sh
```
#!/bin/bash
while true; do
        docker pause testone
        docker unpause testone
done
```
3. run exec.sh
```
#!/bin/bash
while true; do
        echo "exec ${i}th times"
        docker exec -ti testone bash  -c "sleep 0.5 ;exit"
        ((i++))
done
```
just wait a minute, you will see the blocking.

you can also modify daemon'code to help reproduce like this:
```
--- a/api/server/router/container/exec.go
+++ b/api/server/router/container/exec.go
@@ -77,7 +77,7 @@ func (s *containerRouter) postContainerExecStart(ctx context.Context, w http.Res
        if err := json.NewDecoder(r.Body).Decode(execStartCheck); err != nil {
                return err
        }
-
+       time.Sleep(2 * time.Second)
        if exists, err := s.backend.ExecExists(execName); !exists {
                return err
        }
```
then run " docker unpause testone; sleep 2;docker pause testone" in a shell, and run " docker exec -ti testone bash  -c "sleep 0.5 ;exit"" in another shell immediately.

### CODE REVIEW

**Comment by keloyang** [comment]
ping @dnephin

**Comment by coolljt0725** [comment]
ping @vdemeester  @thaJeztah

File: client/hijack.go [hunk 1]
```diff
@@ -72,7 +74,21 @@ func (cli *Client) postHijacked(ctx context.Context, path string, query url.Valu
 	defer clientconn.Close()
 
 	// Server hijacks the connection, error 'connection closed' expected
-	_, err = clientconn.Do(req)
+	resp, err := clientconn.Do(req)
+	if err != nil {
+		return types.HijackedResponse{}, err
+	}
+	if resp != nil {
```
Comment (cpuguy83) [inline client/hijack.go]: This nil check looks like it's not needed.

**Review by cpuguy83 — APPROVED** [review]
LGTM

File: client/hijack.go (hunk 1 above)
Comment (thaJeztah) [inline client/hijack.go]: Yes, I think it's cleaner without @keloyang can you remove it?

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: client/hijack.go [hunk 2]
```diff
@@ -72,7 +74,21 @@ func (cli *Client) postHijacked(ctx context.Context, path string, query url.Valu
...
+		if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusSwitchingProtocols {
+			defer resp.Body.Close()
+
+			body, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				return types.HijackedResponse{}, err
+			}
+			return types.HijackedResponse{}, fmt.Errorf("Error response from daemon: %s", bytes.TrimSpace(body))
+		}
+	}
```
Comment (dnephin) [inline client/hijack.go]: What do you think about something like this:

```go

defer ensureReaderClosed(resp)

switch resp.StatusCode {
case http.StatusOK, http.StatusSwitchingProtocols:
    rwc, br := clientconn.Hijack()
    return types.HijackedResponse{Conn: rwc, Reader: br}, nil
}

body, err := ioutil.ReadAll(resp.Body)
...
return types.HijackedResponse{}, fmt.Errorf("...")
```

I think you want the response closed called in both cases?

File: client/hijack.go (hunk 1 above)
Comment (keloyang) [inline client/hijack.go]: It's true that I need remove it , thank you.

File: client/hijack.go (hunk 2 above)
Comment (keloyang) [inline client/hijack.go]: @dnephin ，thanks for your review and advice. 
I just want the response close called when resp.StatusCode != http.StatusOK(or http.StatusSwitchingProtocols)， If  resp.StatusCode is http.StatusSwitchingProtocols, do we must close the resp ? I'm not sure.

**Comment by keloyang** [comment]
ping @thaJeztah

**Comment by keloyang** [comment]
ping @thaJeztah @dnephin @aaronlehmann

File: client/hijack.go [hunk 3]
```diff
@@ -72,11 +74,23 @@ func (cli *Client) postHijacked(ctx context.Context, path string, query url.Valu
 	defer clientconn.Close()
 
 	// Server hijacks the connection, error 'connection closed' expected
-	_, err = clientconn.Do(req)
-
-	rwc, br := clientconn.Hijack()
+	resp, err := clientconn.Do(req)
+	if resp != nil {
```
Comment (thaJeztah) [inline client/hijack.go]: I think the previous comment https://github.com/moby/moby/pull/33262#discussion_r117484500 meant to say that we don't have to check `resp` if `err != nil`; we can still return early here, to reduce complexity, e.g.;

```go
resp, err := clientconn.Do(req)
if err != nil {
	return types.HijackedResponse{}, err
}

defer resp.Body.Close()

switch resp.StatusCode {
case http.StatusOK, http.StatusSwitchingProtocols:
	rwc, br := clientconn.Hijack()
	return types.HijackedResponse{Conn: rwc, Reader: br}, err
}
body, err := ioutil.ReadAll(resp.Body)
if err != nil {
	return types.HijackedResponse{}, err
}
return types.HijackedResponse{}, fmt.Errorf("Error response from daemon: %s", bytes.TrimSpace(body))
```

**Review by thaJeztah — CHANGES_REQUESTED** [review]
left one suggestion, but LGTM otherwise

ping @dnephin PTAL as well

**Comment by coolljt0725** [comment]
@keloyang  squash your commits

**Comment by keloyang** [comment]
@coolljt0725 , I apologize and update the pr. Thank you.

**Review by dnephin — APPROVED** [review]
LGTM

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

File: client/hijack.go [hunk 4]
```diff
@@ -72,11 +74,23 @@ func (cli *Client) postHijacked(ctx context.Context, path string, query url.Valu
...
-	_, err = clientconn.Do(req)
+	resp, err := clientconn.Do(req)
+	if err != nil {
+		return types.HijackedResponse{}, err
+	}
 
-	rwc, br := clientconn.Hijack()
+	defer resp.Body.Close()
+	switch resp.StatusCode {
+	case http.StatusOK, http.StatusSwitchingProtocols:
```
Comment (tonistiigi) [inline client/hijack.go]: @keloyang Why is `StatusOK` allowed here?

**Comment by keloyang** [comment]
```
func (s *containerRouter) postContainersAttach(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
        err := httputils.ParseForm(r)
        if err != nil {
                return err
        }
        containerName := vars["name"]

        _, upgrade := r.Header["Upgrade"]
        detachKeys := r.FormValue("detachKeys")

        hijacker, ok := w.(http.Hijacker)
        if !ok {
                return fmt.Errorf("error attaching to container %s, hijack connection missing", containerName)
        }

        setupStreams := func() (io.ReadCloser, io.Writer, io.Writer, error) {
                conn, _, err := hijacker.Hijack()
                if err != nil {
                        return nil, nil, nil, err
                }

                // set raw mode
                conn.Write([]byte{})

                if upgrade {
                        fmt.Fprintf(conn, "HTTP/1.1 101 UPGRADED\r\nContent-Type: application/vnd.docker.raw-stream\r\nConnection: Upgrade\r\nUpgrade: tcp\r\n\r\n")
                } else {
                        fmt.Fprintf(conn, "HTTP/1.1 200 OK\r\nContent-Type: application/vnd.docker.raw-stream\r\n\r\n")
                }
...
```

**Comment by tonistiigi** [comment]
@keloyang But this function always sends the `upgrade` header so it should not reach the second condition.

**Comment by keloyang** [comment]
@tonistiigi thanks. I have found that out, if we don't  allowe StatusOK  in postHijack，we should remove 'fmt.Fprintf(conn, "HTTP/1.1 200 OK\r\nContent-Type: application/vnd.docker.raw-stream\r\n\r\n")' from docker daemon, WDYT ?
and ping @dnephin @thaJeztah

**Comment by tonistiigi** [comment]
It may be backward compatibility with older clients(didn't confirm), but shouldn't be a problem for the client pkg as we didn't support new clients with old daemon until very recently.

**Comment by keloyang** [comment]
Do we have any problems to allow  StatusOK in postHijack ?

**Pull request merged**